### PR TITLE
Added Instagram Provider

### PR DESCRIPTION
**Pull request by yannicsmeets**
Includes the option to enforce signed requests according to the documentation of instagram: https://instagram.com/developer/secure-api-requests/

### CODE REVIEW

**Comment by PinpointTownes** [comment]
Thanks for this contribution and sorry for the delay! I'll give it a look ASAP :clap:

File: AspNet.Security.OAuth.Providers.sln.DotSettings [hunk 1]
```diff
@@ -21,4 +21,7 @@
     <s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/SPACE_BEFORE_TYPEOF_PARENTHESES/@EntryValue">False</s:Boolean>
     <s:String x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/TYPE_DECLARATION_BRACES/@EntryValue">END_OF_LINE</s:String>
     <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateInstanceFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String>
-    <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String></wpf:ResourceDictionary>
\ No newline at end of file
+    <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String>
```
Comment (PinpointTownes) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: Unrelated changes, please revert them.

File: samples/Mvc.Client/project.json [hunk 2]
```diff
@@ -50,7 +51,7 @@
 
     "commands": {
         "kestrel": "Microsoft.AspNet.Hosting --server=Microsoft.AspNet.Server.Kestrel --server.urls=http://localhost:54540",
-        "web": "Microsoft.AspNet.Hosting server=Microsoft.AspNet.Server.WebListener server.urls=http://localhost:54540"
+        "web": "Microsoft.AspNet.Server.Kestrel"
```
Comment (PinpointTownes) [inline samples/Mvc.Client/project.json]: Revert that too, please :smile:

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationDefaults.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpoint = "https://api.instagram.com/oauth/access_token";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+        /// </summary>
+        public const string UserInformationEndpoint = "/users/self";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationDefaults.cs]: Please use an absolute URL here, to match the other endpoints.

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationDefaults.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+        /// <summary>
+        /// Default location of the api
+        /// </summary>
+        public const string ApiLocation = "https://api.instagram.com/v1";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationDefaults.cs]: This const won't be necessary when `UserInformationEndpoint` is updated to use an absolute URL :clap:

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+namespace AspNet.Security.OAuth.Instagram {
+    public class InstagramAuthenticationHandler : OAuthHandler<InstagramAuthenticationOptions> {
+        public InstagramAuthenticationHandler([NotNull] HttpClient client)
+            : base(client) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var userInformationEndpoint = $"{Options.UserInformationEndpoint}?access_token={tokens.AccessToken}";
+
+            if (Options.SignedRequestsEnforced) {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: I'm not opposed to implement this (non-standard) feature, but TBH, it's almost unnecessary in our case, since access tokens will often (well, always :smile:) be stored in encrypted cookies (and not as-is in a database).

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                var parameters = new Dictionary<string, string> {
+                    ["access_token"] = tokens.AccessToken
+                };
+
+                var sig = SignRequest(InstagramAuthenticationDefaults.UserInformationEndpoint,
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: `signature`

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        public InstagramAuthenticationHandler([NotNull] HttpClient client)
+            : base(client) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var userInformationEndpoint = $"{Options.UserInformationEndpoint}?access_token={tokens.AccessToken}";
+
+            if (Options.SignedRequestsEnforced) {
+                var parameters = new Dictionary<string, string> {
+                    ["access_token"] = tokens.AccessToken
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: If `access_token` is always the only parameter, is it worth using a dictionary? It may help simplifying your signature generation path.

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        public InstagramAuthenticationOptions() {
+            AuthenticationScheme = InstagramAuthenticationDefaults.AuthenticationScheme;
+            DisplayName = InstagramAuthenticationDefaults.DisplayName;
+            ClaimsIssuer = InstagramAuthenticationDefaults.Issuer;
+
+            CallbackPath = new PathString(InstagramAuthenticationDefaults.CallbackPath);
+
+            AuthorizationEndpoint = InstagramAuthenticationDefaults.AuthorizationEndpoint;
+            TokenEndpoint = InstagramAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = InstagramAuthenticationDefaults.ApiLocation +
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs]: Don't forget to remove `InstagramAuthenticationDefaults.ApiLocation`.

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            return new AuthenticationTicket(context.Principal, context.Properties, context.Options.AuthenticationScheme);
+        }
+
+        private static byte[] CreateToken([NotNull] string endpoint, [NotNull] Dictionary<string, string> parameters) {
+            var token = parameters.OrderBy(par => par.Key)
+                .Aggregate(endpoint, (current, par) => current + $"|{par.Key}={par.Value}");
+            return Encoding.UTF8.GetBytes(token);
+        }
+
+        private static string ConvertToHex(byte[] hash) {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Consider merging this method (and the previous one) with `SignRequest`.

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs (hunk 5 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Not to mention that this option will only impact the userinfo request, which is made before the access token is persisted some (it's still in memory at this point), so the security gain is really marginal.

**Comment by PinpointTownes** [comment]
Minor my few remarks, it looks fine :+1:

**Comment by yannicsmeets** [comment]
Thanks for the feedback!

I added the SignedRequestEnforced as that is an option that can be set for an application within Instagram. If an application uses the instagram api for more than just oauth, it could be that this option if set. If so, this signature is required for all requests, also the ones required for oauth.

File: AspNet.Security.OAuth.Providers.sln.DotSettings [hunk 10]
```diff
@@ -21,4 +21,6 @@
     <s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/SPACE_BEFORE_TYPEOF_PARENTHESES/@EntryValue">False</s:Boolean>
     <s:String x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/TYPE_DECLARATION_BRACES/@EntryValue">END_OF_LINE</s:String>
     <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateInstanceFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String>
-    <s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String></wpf:ResourceDictionary>
\ No newline at end of file
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EAddAccessorOwnerDeclarationBracesMigration/@EntryIndexedValue">True</s:Boolean>
```
Comment (PinpointTownes) [inline AspNet.Security.OAuth.Providers.sln.DotSettings]: Note for me: remove that before merging :smile:

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+using Microsoft.AspNet.Authentication.OAuth;
+using Microsoft.AspNet.Http;
+
+namespace AspNet.Security.OAuth.Instagram {
+    /// <summary>
+    /// Defines a set of options used by <see cref="InstagramAuthenticationHandler"/>.
+    /// </summary>
+    public class InstagramAuthenticationOptions : OAuthOptions {
+        public bool SignedRequestsEnforced { get; set; } = false;
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs]: Note for me: remove the explicit `false` before merging (since it's the default value) and add some documentation.

**Comment by PinpointTownes** [comment]
I'm sold :smile: 

(I'll probably change the name before merging. How about `UseSignedRequests`?)

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs (hunk 11 above)
Comment (yannicsmeets) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs]: Should I not give it default value?

**Comment by yannicsmeets** [comment]
That sounds good! Shall I squash it in one commit?

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs (hunk 11 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs]: `false` is the default value for `bool` variables/fields/properties you don't explicitly set, so it's not really necessary.

**Comment by PinpointTownes** [comment]
> That sounds good! Shall I squash it in one commit?

Yep, thanks! (if you want to fix the remarks I added before squashing, don't hesitate) :smile:

**Comment by yannicsmeets** [comment]
Obviously ;) Do these look good?

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Instagram {
+    public class InstagramAuthenticationHandler : OAuthHandler<InstagramAuthenticationOptions> {
+        public InstagramAuthenticationHandler([NotNull] HttpClient client)
+            : base(client) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var userInformationEndpoint = $"{Options.UserInformationEndpoint}?access_token={tokens.AccessToken}";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Wooops, I missed this one: please use `QueryHelpers` instead of manually concatenating the parameters :smile:

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+            if (Options.UseSignedRequests) {
+                var signature = SignRequest("/users/self",
+                    tokens.AccessToken, Options.ClientSecret);
+
+                userInformationEndpoint += $"&sig={signature}";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Same remark :+1:

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    public class InstagramAuthenticationHandler : OAuthHandler<InstagramAuthenticationOptions> {
+        public InstagramAuthenticationHandler([NotNull] HttpClient client)
+            : base(client) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var userInformationEndpoint = $"{Options.UserInformationEndpoint}?access_token={tokens.AccessToken}";
+
+            if (Options.UseSignedRequests) {
+                var signature = SignRequest("/users/self",
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: I'm not a huge fan of this hardcoded path (specially since it's essentially `userInformationEndpoint` without the scheme, domain and base path) but I can't find a better alternative, so let's keep it as-is :smile:

Comment (yannicsmeets) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Yeah, thats why I initially put it in the Defaults class. Even though that isn't pretty either...

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+            CallbackPath = new PathString(InstagramAuthenticationDefaults.CallbackPath);
+
+            AuthorizationEndpoint = InstagramAuthenticationDefaults.AuthorizationEndpoint;
+            TokenEndpoint = InstagramAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = InstagramAuthenticationDefaults.UserInformationEndpoint;
+
+            Scope.Add("basic");
+
+            SaveTokensAsClaims = false;
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationOptions.cs]: Note for me: remove before merging (https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/5d5b6ec7fdb338b7c72842f0f83ea0cca051c353)

File: src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            if (context.Principal?.Identity == null) {
+                return null;
+            }
+
+            return new AuthenticationTicket(context.Principal, context.Properties, context.Options.AuthenticationScheme);
+        }
+
+        private static string SignRequest([NotNull] string endpoint, [NotNull] string accessToken,
+            string secret) {
+            var token = Encoding.UTF8.GetBytes($"{endpoint}|access_token={accessToken}");
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Hmmmm, the way the signature is formatted is not really foolproof and won't work if the user adds query parameters in the userinfo URL.

I hacked a different version. Could you please give it a try? (I don't have an Instagram account :smile:)

```csharp
protected virtual string ComputeSignature(string address) {
    using (var algorithm = new HMACSHA256(Encoding.UTF8.GetBytes(Options.ClientSecret))) {
        // Extract the parameters from the query string.
        var parameters = from parameter in QueryHelpers.ParseQuery(address)
                         orderby parameter.Key
                         from value in parameter.Value
                         orderby value
                         select $"{parameter.Key}={value}";

        // See https://www.instagram.com/developer/secure-api-requests/
        // for more information about the signature format.
        var bytes = Encoding.UTF8.GetBytes($"/users/self|{string.Join("|", parameters)}");

        // Compute the HMAC256 signature.
        var hash = algorithm.ComputeHash(bytes);

        // Convert the hash to its lowercased hexadecimal representation.
        return BitConverter.ToString(hash).Replace("-", "").ToLower();
    }
}
```

Comment (yannicsmeets) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: I checked this, and there are some issues with it. 
* The QueryHelpers.ParseQuery doesn't work as I expected. It seems to parse the url up until the first parameter as the first parameter. So http://localhost:54540/signin?key=value will get parsed to {http://localhost:54540/signin?key, value}. Seems like a bug to me.
* The values of the parameters are all strings. Doing `from value in parameter.Value orderby value`will therefore result in adding a pair for each character in value. I think those two lines can be omitted. 

Finally a question for my curiosity :smile:, why make this method protected? Are other classes in this namespace going to use this method?

Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: > The QueryHelpers.ParseQuery doesn't work as I expected. It seems to parse the url up until the first parameter as the first parameter. So http://localhost:54540/signin?key=value will get parsed to {http://localhost:54540/signin?key, value}. Seems like a bug to me.

Ah yeah, my fault. Sounds like it doesn't accept a full URL, but just the query string. I guess `IndexOf` + `Substring` would do the trick. Do you wanna try?

> The values of the parameters are all strings. Doing from value in parameter.Value orderby valuewill therefore result in adding a pair for each character in value. I think those two lines can be omitted.

Hahaha, the downsides of implicit conversions :smile: 
I guess a `StringValues.ToArray()` call would work, but there's probably no point supporting duplicate values.

> Finally a question for my curiosity :smile:, why make this method protected? Are other classes in this namespace going to use this method?

Nope, but developers are free to subclass the handler class to override the default logic. Probably just a matter of personal preference.

Comment (yannicsmeets) [inline src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs]: Thanks for the info! :smile: I will try to fix the parameter fetching.

**Comment by PinpointTownes** [comment]
Merged, thanks for your contribution! :tada: 

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/5e7740c5237245ef14d7e014787110c3b945c143
